src/resource/ubridge.c

```diff
@@ -512,6 +512,59 @@ static const char *_get_iov_str(struct buffer *buf, bool unset, struct iovec *io
 	return NULL;
 }
 
+static int _write_kv_store_dump(struct buffer *buf, sid_resource_t *kv_store_res)
+{
+	kv_store_iter_t *iter;
+	const char *key = "<NO KEY>";
+	struct usid_dump_header header;
+	size_t size;
+	kv_store_value_flags_t flags;
+	void *value;
+	struct iovec tmp_iov[KV_VALUE_IDX_DATA + 1];
+	struct iovec *iov;
+	unsigned int i, r = 0;
+	uint32_t len;
+	void *data;
+
+	if (!(iter = kv_store_iter_create(kv_store_res))) {
+		log_error(ID(kv_store_res), INTERNAL_ERROR "%s: failed to create record iterator", __func__);
+		return -ENOMEM;
+	}
+
+	while ((value = kv_store_iter_next(iter, &size, &flags))) {
+		key = kv_store_iter_current_key(iter);
+		if (!strncmp(key, "U:", 2))
+			continue;
+		iov = _get_value_vector(flags, value, size, tmp_iov);
+		header.seqnum = KV_VALUE_SEQNUM(iov);
+		header.flags = KV_VALUE_FLAGS(iov);
+		header.data_count = (flags & KV_STORE_VALUE_VECTOR)? size - KV_VALUE_IDX_DATA : 1;
+		if (!buffer_add(buf, &header, sizeof(header), &r))
+			goto out;
+		len = strlen(key) + 1;
+		if (!buffer_add(buf, &len, sizeof(len), &r))
+			goto out;
+		if (!buffer_add(buf, (void *)key, len, &r))
+			goto out;
+		for (i = 0; i < header.data_count + 1; i++) {
+			data = iov[KV_VALUE_IDX_OWNER + i].iov_base;
+			len = iov[KV_VALUE_IDX_OWNER + i].iov_len;
+			if (!buffer_add(buf, &len, sizeof(len), &r))
+				goto out;
+			if (!buffer_add(buf, data, len, &r))
+				goto out;
+		}
+	}
+	/* add zeroed header to signal the end of the dump */
+	memset(&header, 0, sizeof(header));
+	buffer_add(buf, &header, sizeof(header), &r);
```

Comment:
This is probably not needed? We also get the size of the message on the other end (`sidctl`), so we should already know when to end without a need to check for the trailing zeroed header.